### ISSUE TITLE
Insertion to an outbox queue stops unless it is completely drained when HighWaterMark is reached.

### ISSUE THREAD

**Issue by bipulkarnani** [hazelcast/hazelcast-jet#601]
My understanding is if an outbox queue is full, then insertion stops unless outbox queue is drained out completely. 

If above implementation is right then it has consequences where if source is `Kafka`, time between two consecutive poll may exceed `max.poll.interval.ms`, leading to `commitFailedException` error.

**Comment by viliam-durina** [comment]
It's partially true for Jet 0.4: if the outbox queue is full, insertion stops unless at least one item is removed from the outbox by downstream processor (not until it's drained completely).

For the Kafka source, see #428. In short: yes, we might have exceeded the `max.poll.interval.ms`, but it is fixed in 0.5 where we use manual partition assignment. 0.5 should be out in the matter of days.

**Comment by viliam-durina** [comment]
Duplicate of #428

**Comment by bipulkarnani** [comment]
Hi @viliam-durina ,

Thanks for the reply. I am waiting for 0.5. 

In the meanwhile to better understand the Jet internals, me and another team member did small experiment with the `edge queue size` of the processor and here is the output of it:

```
I have created a sample dag(jet 0.4) with two vertices StreamKafkaP source and Logger sink with an `edge queue size=20` and kafka `max.poll.records=2` and `localParallelism=1` for both vertices and before starting jet I already had 100 simple messages in the topic subscribed by StreamKafkaP vertex. I made the Logger vertex to sleep after printing a logger message and before returning so that I can check the current offset of the kafka consumer.

In the above scenario I have observed `CURRENT-OFFSET=34` when first message has been logged and `CURRENT-OFFSET=66` when second message is logged instead of it being around 34 + 2(since max.poll.records=2 and queue size is already full). The CURRENT-OFFSET remains stagnant till around 30 messages are processed and then CURRENT-OFFSET becomes 98.

Ideally offset should increase by 2 (i.e max.poll.records) after sleep period of processor is over for two messages since localparallelism is 1, instead it got increased by more than queue size (i.e. 32).
```
It will be really helpful if you can go through the above experiment and explain what we are missing in the above experiment.

**Comment by viliam-durina** [comment]
Jet 0.5 is out in the meantime. 

The edge queue size is the maximum number of objects waiting in queue. There is one queue between each processor instance and the number of processor instances on each member is determined by local parallelism of your member (local parallelism is equal to number of cpus in 0.4 and is 2 in 0.5 for sources). I don't know your parallelism. Also if the edge after the kafka processor is partitioned, each item has to go to specific downstream processor.

Kafka processor works like this: it polls kafka for records (at most for 2 records in your case). Then it tries to add those records to queues, which have capacity of 20 in your case and there can be multiple of them. If the queue is full, it stops and tries again later.

Knowing this, the numbers in your example look possible. Can you please share the DAG code to be sure?

**Comment by bipulkarnani** [comment]
> Kafka processor works like this: it polls kafka for records (at most for 2 records in your case). Then it tries to add those records to queues, which have capacity of 20 in your case and there can be multiple of them. If the queue is full, it stops and tries again later.

Given that `local parallelism` = 1 for both kafka source and logger vertex, by the above logic kafka offset should increase by 2 (since we put sleep in the logger vertex) but we observed that it increased by 32.

**Comment by viliam-durina** [comment]
Was it also 1 jet member? Then it probably should be 22 (20+2) after first event and 24 after second and third and so forth. How do you check the CURRENT-OFFSET? Again, showing us the code will help a lot.

**Comment by bipulkarnani** [comment]
> Was it 1 jet member?
Yes it is one jet member.

> How do you check the CURRENT-OFFSET? 
kafka-consumer-groups.sh given in the kafka distribution.

> Code to test the same is: https://gist.github.com/bipulkarnani/c2cb5f66e525a2ddaa29c035a55435ca

**Comment by viliam-durina** [comment]
I investigated it and it's little more complex:
1. the queue size is rounded up to next power of 2, in your case to 32. This is for performance reasons.
2. the way the buffers are drained: the target processor always drains whatever there is in the queue and copies it to local buffer to avoid concurrent reads.

This explains what you've observed:
- after first item the offset was 34: that is 32 (real queue size) and 2 (the polled items waiting to be inserted to queue)
- since the max. poll size is 2, initially two items are put into the queue and those two are read by sink into local buffer
- after third item (not second), 32 items are drained from queue to local buffer, which allows the source to enqueue another 32 items, thus the observed offset value was 66.

Thanks for spotting this out, we'll document it better in `EdgeConfig`.

**Comment by bipulkarnani** [comment]
@viliam-durina: Thank you for the explanation.
